### DISCUSSION TITLE
Template

### DISCUSSION THREAD

**Discussion by rj10rd** [arendst/Tasmota#14847] [Q&A] [answered]
Sir I want to add my custom sonoff basic r2 template
When compling custom bin
How i can achieve this.

**Comment by sfromis** [comment, marked as answer]
By setting values in `user_config_override.h`, you can find the most common changes in [my_user_config.h](https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L52-L61) (link points at relevant section with the two #define for configuring and selecting template).

**Reply by rj10rd** [reply]
i successfully  added my custom template using define USER_TEMPLATE
but please help me how to set my own custom template as default

**Reply by ascillato** [reply]
To set a custom module at compile time, first disable the actual default and then add your custom template as follows in **_user_config_override.h_** file:

```cpp
// Setting Default Modules. GPIOs template for devices (Reference https://templates.blakadder.com/)
// -----------------------

#ifdef MODULE
#undef MODULE
#endif
#define MODULE                 USER_MODULE      // [Module] Select default model (the list is kModuleNiceList() in file tasmota_template.h)

#ifdef FALLBACK_MODULE
#undef FALLBACK_MODULE
#endif
#define FALLBACK_MODULE        USER_MODULE      // to Select the default model as FALLBACK when the user does a RESET 1 

#ifdef USER_TEMPLATE
#undef USER_TEMPLATE
#endif
#define USER_TEMPLATE    "{\"NAME\":\"Shelly 1\",\"GPIO\":[0,0,0,0,224,192,0,0,0,0,0,0,0,4736],\"FLAG\":0,\"BASE\":46}"  // Shelly 1 Example Template
```

If besides that, you want to add rules and you want them to be enabled at compile time:
```cpp
#define USER_RULE1 "ON System#Boot DO Var1 1 ENDON"
#define USER_BACKLOG "Rule1 1"
```

**Reply by rj10rd** [reply]
> To set a custom module at compile time, first disable the actual default and then add your custom template as follows in **_user_config_override.h_** file:
> 
> ```c++
> // Setting Default Modules. GPIOs template for devices (Reference https://templates.blakadder.com/)
> // -----------------------
> 
> #ifdef MODULE
> #undef MODULE
> #endif
> #define MODULE                 USER_MODULE      // [Module] Select default model (the list is kModuleNiceList() in file tasmota_template.h)
> 
> #ifdef FALLBACK_MODULE
> #undef FALLBACK_MODULE
> #endif
> #define FALLBACK_MODULE        USER_MODULE      // to Select the default model as FALLBACK when the user does a RESET 1 
> 
> #ifdef USER_TEMPLATE
> #undef USER_TEMPLATE
> #endif
> #define USER_TEMPLATE    "{\"NAME\":\"Shelly 1\",\"GPIO\":[0,0,0,0,224,192,0,0,0,0,0,0,0,4736],\"FLAG\":0,\"BASE\":46}"  // Shelly 1 Example Template
> ```
> 
> If besides that, you want to add rules and you want them to be enabled at compile time:
> 
> ```c++
> #define USER_RULE1 "ON System#Boot DO Var1 1 ENDON"
> #define USER_BACKLOG "Rule1 1"
> ```

thank you sir for saving my night

**Reply by alexdelprete** [reply]
I'm trying to activate the template following your recommendation, but it's not working.

this is the relevant section:

```cpp
// [Template] Set JSON user template for GPIO configuration (Atom+Tail and ESP8266)

// Define Template based on chip
#ifdef USER_TEMPLATE
  #undef USER_TEMPLATE
#endif

#ifdef ESP32
  #define USER_TEMPLATE "{\"NAME\":\"NeoPool Atom Lite + Tail485\",\"GPIO\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,1,1,0,0,6976,0,0,0,0,0,7008,1,0,0,0,0,0,0],\"FLAG\":0,\"BASE\":1}"
#endif  // ESP32

#ifdef ESP8266
  #warning **** Template: NeoPool ESP8266 ****
  #define USER_TEMPLATE "{\"NAME\":\"NeoPool ESP8266\",\"GPIO\":[1,6528,1,6496,1,1,0,0,1,1,1,1,1,1],\"FLAG\":0,\"BASE\":18}"
#endif  // ESP8266

// Activate User Template
#ifdef MODULE
  #undef MODULE
#endif
#define MODULE                 USER_MODULE  // [Module] Select default model (the list is kModuleNiceList() in file tasmota_template.h) USER_MODULE is the TEMPLATE

#ifdef FALLBACK_MODULE
  #undef FALLBACK_MODULE
#endif
#define FALLBACK_MODULE        USER_MODULE   // to Select the default model as FALLBACK when the user does a RESET 1
```

But when I flash and boot the device, Module 1 is active, not Module 0.

In the end I solved this way, but I wonder if it's appropriate:

```cpp
#undef USER_BACKLOG
#define USER_BACKLOG "Rule1 4; Rule1 1; Module 0;"
```

**Reply by curzon01** [reply]
on your test device you must reset the settings to default to apply the changes, because you already have a valid config:
WebGUI: _Configuraton -> Reset Configuration_ or Console e. g. `Reset 4`

not needed on fresh ESP using this setup - for test: erase ESP and flash your customized FW

**Reply by sfromis** [reply]
Can't see what happened, but `#define MODULE USER_MODULE` **will** activate the template, as long as you make sure that Tasmota initializes fully after flashing, instead of continuing on an old config. And yes, I've tried that and seen it work.

Your `USER_BACKLOG` does not do much without you also having `USER_RULE1` in the build, to have a rule to activate.

**Reply by curzon01** [reply]
> as long as you make sure that Tasmota initializes fully after flashing

that's I wrote, or I'm wrong?

**Reply by alexdelprete** [reply]
> on your test device you must reset the settings to default to apply the changes, because you already have a valid config

I'm always erasing the flash before reflashing. So I guess I don't need to reset.

**Reply by curzon01** [reply]
> I'm always erasing the flash before reflashing. So I guess I don't need to reset.

I don't know how you do this but it works on a fresh ESP or after the reset to firmwarte defaults, try it...

**Reply by alexdelprete** [reply]
> as long as you make sure that Tasmota initializes fully after flashing, instead of continuing on an old config

you mean a fully reset device? I erase flash with the flashing tool of M5Stack before flashing the bin. So it is fully reset. But the template doesn't activate.

> Your `USER_BACKLOG` does not do much without you also having `USER_RULE1` in the build, to have a rule to activate.

I didn't provide the entire content of the file with the `USER_RULE1`, just the snippet to show how I solved the template activation by adding `Module 0` to the `USER_BACKLOG`.

This is the whole file, and it is working (tested on 2 new Atom Lite devices), but I had to add `Module 0` to make it activate the template.

```cpp
// PROJECT is used as the default topic delimiter
#undef PROJECT
#define PROJECT                "DevSmartPool"

// [Template] Set JSON user template for GPIO configuration (Atom+Tail and ESP8266)

// Define Template based on chip
#ifdef USER_TEMPLATE
  #undef USER_TEMPLATE
#endif

#ifdef ESP32
  #define USER_TEMPLATE "{\"NAME\":\"NeoPool Atom Lite + Tail485\",\"GPIO\":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,1,1,1,0,0,6976,0,0,0,0,0,7008,1,0,0,0,0,0,0],\"FLAG\":0,\"BASE\":1}"
#endif  // ESP32

#ifdef ESP8266
  #define USER_TEMPLATE "{\"NAME\":\"NeoPool ESP8266\",\"GPIO\":[1,6528,1,6496,1,1,0,0,1,1,1,1,1,1],\"FLAG\":0,\"BASE\":18}"
#endif  // ESP8266

// Activate User Template
#ifdef MODULE
  #undef MODULE
#endif
#define MODULE                 USER_MODULE  // [Module] Select default model (the list is kModuleNiceList() in file tasmota_template.h) USER_MODULE is the TEMPLATE

#ifdef FALLBACK_MODULE
  #undef FALLBACK_MODULE
#endif
#define FALLBACK_MODULE        USER_MODULE   // to Select the default model as FALLBACK when the user does a RESET 1

// [DeviceName] Devicename up to 32 characters
#undef DEVICE_NAME
#define DEVICE_NAME            PROJECT

// [FriendlyName] Friendlyname up to 32 characters used by webpages and Alexa
#undef FRIENDLY_NAME
#define FRIENDLY_NAME          PROJECT

// [Topic] unique MQTT device topic
#undef MQTT_TOPIC
#define MQTT_TOPIC             PROJECT

// [MqttClient] Also fall back topic using last 6 characters of MAC address or use "DVES_%12X" for complete MAC address
#undef MQTT_CLIENT_ID
#define MQTT_CLIENT_ID         PROJECT

// [SensorRetain] Sensor may send retain flag (false = off, true = on)
#undef MQTT_SENSOR_RETAIN
#define MQTT_SENSOR_RETAIN     true

// Add support for rules (+8k code)
#undef USE_RULES
#define USE_RULES

// Add Rule1 data saved at initial firmware load or when command reset is executed
#undef USER_RULE1
#define USER_RULE1 "ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"

// Activate user Rule1, set WifiPower to dynamic and disable any device discovery protocol
#undef USER_BACKLOG
#define USER_BACKLOG "Rule1 4; Rule1 1; WifiPower 0; Module 0;"

// Log modbus traffic errors: use Weblog 3 or 4 to debug and Weblog 2 to go back to normal log level
// https://github.com/arendst/Tasmota/discussions/18991#discussioncomment-6317229
// #undef DEBUG_TASMOTA_SENSOR
// #define DEBUG_TASMOTA_SENSOR

// Add support for NeoPool module
#undef USE_NEOPOOL
#define USE_NEOPOOL

#endif  // _USER_CONFIG_OVERRIDE_H_
```

**Reply by alexdelprete** [reply]
> I don't know how you do this but it works on a fresh ESP or after the reset to firmwarte defaults, try it...

I do it with the M5Stack native tool. But I can try with esptool too.

Anyway, I'll try with your recommended command `Reset 4` now. Will give you a feedback.

**Reply by alexdelprete** [reply]
Is PlatformIO vscode plugin reliable for erasing the flash? I'm using that to edit and compile/flash now.

![image](https://github.com/arendst/Tasmota/assets/7027842/f183399e-9f0e-4fce-a25b-18c1f5a77bd5)

**Reply by sfromis** [reply]
In case of boot loops, Tasmota may try to recover by "scaling back" features until able to boot without quickly crashing again. One step of this process is to deactivate the template. As causes of a boot loop may not be entirely predictable, what you see may be a question of "luck" with the recovery not happening the same way when you used a different build.

**Reply by curzon01** [reply]
> Is PlatformIO vscode plugin reliable for erasing the flash? I'm using that to edit and compile/flash now.

I'm unsure, using esptool it works:

` esptool.py --port /dev/ttyUSB0 --baud 115200 --chip esp32 erase_flash`

> I'll try with your recommended command Reset 4 now

better suggest a [reset 5](https://tasmota.github.io/docs/Commands/#reset), this additonal clears the flash before but keeps wifi settings

**Reply by alexdelprete** [reply]
> better suggest a [reset 5](https://tasmota.github.io/docs/Commands/#reset), this additonal clears the flash before but keeps wifi settings

Found out PlatformIO uses esptool to flash/erase etc. So the procedure was correct. I also tried `Reset 5` and `Reset 6`.

Anyway, I now tried the manual procedure, after compilation (no `Module 0` in `USER_BACKLOG`).

_Erased flash with esptool:_

```console
> esptool erase_flash

esptool.py v4.6.2
Found 1 serial ports
Serial port COM3
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting......
Detecting chip type... ESP32
Chip is ESP32-PICO-D4 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, Embedded Flash, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: e8:9f:6d:09:63:c0
Uploading stub...
Running stub...
Stub running...
Erasing flash (this may take a while)...
Chip erase completed successfully in 10.2s
Hard resetting via RTS pin...
```

_Flashed the bin:_

```console
> esptool -b 1500000 write_flash 0x0 firmware.factory.bin

esptool.py v4.6.2
Found 1 serial ports
Serial port COM3
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting.........
Detecting chip type... ESP32
Chip is ESP32-PICO-D4 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, Embedded Flash, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: e8:9f:6d:09:63:c0
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 1500000
Changed.
Configuring flash size...
Flash will be erased from 0x00000000 to 0x00341fff...
Compressed 3412880 bytes to 2148517...
Wrote 3412880 bytes (2148517 compressed) at 0x00000000 in 32.6 seconds (effective 836.7 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

_Result is the same: template is not activated on first boot_

```cpp
00:00:00.477 Project DevSmartPool - DevSmartPool Version 13.2.0.2(core2)-2_0_14(2023-11-15T16:15:00)
00:00:00.480 TFS: File '.snsset083' not found
00:00:00.504 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":128,"Free":383,"Rules":"ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"}}
00:00:00.605 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":128,"Free":383,"Rules":"ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"}}
00:00:00.635 RSL: RESULT = {"WifiPower":0.0}
00:00:08.005 WIF: Connecting to AP1 NetStumbling? Channel 1 BSSId 2C:91:AB:AB:72:ED in mode 11n as DevSmartPool-0960...
00:00:09.475 WIF: Connected
00:00:09.729 HTP: Web server active on DevSmartPool-0960 with IP address 10.1.10.168
17:33:00.054 RUL: TIME#INITIALIZED performs "NPTime 0"
17:33:00.058 RSL: RESULT = {"Command":"Unknown"}
17:33:00.701 MQT: Attempting connection...
17:33:00.741 MQT: Connected
17:33:00.744 MQT: tele/DevSmartPool/LWT = Online (retained)
17:33:00.746 MQT: cmnd/DevSmartPool/POWER = 
17:33:00.750 MQT: tele/DevSmartPool/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.2.0.2(core2)","FallbackTopic":"cmnd/DevSmartPool_fb/","GroupTopic":"cmnd/tasmotas/"}}
17:33:00.765 MQT: tele/DevSmartPool/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"DevSmartPool-0960","IPAddress":"10.1.10.168","IP6Global":"","IP6Local":"fe80::ea9f:6dff:fe09:63c0%st2"}}
17:33:00.780 MQT: tele/DevSmartPool/INFO3 = {"Info3":{"RestartReason":"Vbat power on reset","BootCount":2}}
17:33:00.809 RUL: SYSTEM#BOOT performs "Backlog Discover 0; NPTeleperiod 60;"
17:33:00.855 MQT: stat/DevSmartPool/RESULT = {"Discover":"OFF"}
17:33:01.107 MQT: stat/DevSmartPool/RESULT = {"Command":"Unknown"}
```

**Reply by alexdelprete** [reply]
I re-added `Module 0` to `USER_BACKLOG` again, and let PlatformIO recompile/erase/flash automatically:

_First boot: Template is active._

```cpp
00:00:00.001 HDW: ESP32-PICO-D4 v1.0 
00:00:00.023 UFS: FlashFS mounted with 308 kB free
00:00:00.036 CFG: Loaded from File, Count 3
00:00:00.037 FRC: Some settings have been reset (2)
00:00:00.350 BRY: Berry initialized, RAM used 6338 bytes
00:00:00.360 TFS: File '.drvset003' not found
00:00:00.364 Project DevSmartPool - DevSmartPool Version 13.2.0.2(core2)-2_0_14(2023-11-15T17:50:15)
00:00:07.817 WIF: Connecting to AP1 NetStumbling? Channel 1 BSSId 2C:91:AB:AB:72:ED in mode 11n as DevSmartPool-0960...
00:00:09.276 WIF: Connected
00:00:09.533 HTP: Web server active on DevSmartPool-0960 with IP address 10.1.10.168
17:55:28.054 RUL: TIME#INITIALIZED performs "NPTime 0"
17:55:28.358 RSL: RESULT = {"NPTime":"Error"}
17:55:28.695 MQT: Attempting connection...
17:55:28.729 MQT: Connected
17:55:28.731 MQT: tele/DevSmartPool/LWT = Online (retained)
17:55:28.733 MQT: cmnd/DevSmartPool/POWER = 
17:55:28.736 MQT: tele/DevSmartPool/INFO1 = {"Info1":{"Module":"NeoPool Atom Lite + Tail485","Version":"13.2.0.2(core2)","FallbackTopic":"cmnd/DevSmartPool_fb/","GroupTopic":"cmnd/tasmotas/"}}
17:55:28.751 MQT: tele/DevSmartPool/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"DevSmartPool-0960","IPAddress":"10.1.10.168","IP6Global":"","IP6Local":"fe80::ea9f:6dff:fe09:63c0%st2"}}
17:55:28.776 MQT: tele/DevSmartPool/INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":2}}
17:55:28.794 RUL: SYSTEM#BOOT performs "Backlog Discover 0; NPTeleperiod 60;"
17:55:28.840 MQT: stat/DevSmartPool/RESULT = {"Discover":"OFF"}
17:55:29.183 MQT: stat/DevSmartPool/RESULT = {"NPTelePeriod":60}
```

**Reply by alexdelprete** [reply]
I guess I'll keep `Module 0` in `USER_BACKLOG`, it doesn't hurt, and it solves the problem for me (and potentially for other users who might incur in my same issue).

I was afraid of potential bootloops due tue Module command restarting the device, but from what I read, there's no real risk:

![image](https://github.com/arendst/Tasmota/assets/7027842/d2fad01b-2d23-4754-ba84-26bd056a27b8)

**Reply by sfromis** [reply]
That does not show if there was a boot loop causing restoration, which could explain what you see.

**Reply by alexdelprete** [reply]
> That does not show if there was a boot loop causing restoration, which could explain what you see.

I tested without `Module 0`, what should have caused the bootloop?

Furthermore, right after the flashing, I see the console immediately active (2s after) and I see no signs of bootloops.

I could monitor via serial...but I don't think there are bootloops.

**Reply by sfromis** [reply]
You have something wonky going on with your test where you see the template being active.
`FRC: Some settings have been reset (2)` 
That **is** a clear sign of Tasmota detecting a boot loop. Maybe due to `USER_BACKLOG` triggering restarts.

Anyway, I can't see what happens, but I *know* that in a regular working setup, it works fine with `#define MODULE USER_MODULE` without workarounds. Thus something unusual is happening for you, not something general with Tasmota.

**Reply by alexdelprete** [reply]
> That **is** a clear sign of Tasmota detecting a boot loop. Maybe due to `USER_BACKLOG` triggering restarts.

Ok, I removed it. But with `#define MODULE USER_MODULE` it doesn't activate the template, and I don't know how to properly debug why it doesn't.

Here's the log after I flash without `Module 0` in `USER_BACKLOG`, looks clean, but template is not active:

```cpp
00:00:00.074 RUL: Stored uncompressed, would compress from 128 to 81 (-37%)
00:00:00.075 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":128,"Free":383,"Rules":"ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"}}
00:00:00.167 QPC: Reset
00:00:00.465 BRY: Berry initialized, RAM used 6338 bytes
00:00:00.473 TFS: File '.drvset003' not found
00:00:00.478 Project DevSmartPool - DevSmartPool Version 13.2.0.2(core2)-2_0_14(2023-11-15T21:42:48)
00:00:00.481 TFS: File '.snsset083' not found
00:00:00.505 RSL: RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":128,"Free":383,"Rules":"ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"}}
00:00:00.608 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":128,"Free":383,"Rules":"ON Time#Initialized DO NPTime 0 ENDON ON Time#Set DO NPTime 0 ENDON ON System#Boot DO Backlog Discover 0; NPTeleperiod 60; ENDON"}}
00:00:00.638 RSL: RESULT = {"WifiPower":0.0}
00:00:07.964 WIF: Connecting to AP1 NetStumbling? Channel 1 BSSId 2C:91:AB:AB:72:ED in mode 11n as DevSmartPool-0960...
00:00:09.435 WIF: Connected
00:00:09.689 HTP: Web server active on DevSmartPool-0960 with IP address 10.1.10.168
21:46:52.054 RUL: TIME#INITIALIZED performs "NPTime 0"
21:46:52.059 RSL: RESULT = {"Command":"Unknown"}
21:46:52.669 MQT: Attempting connection...
21:46:52.699 MQT: Connected
21:46:52.702 MQT: tele/DevSmartPool/LWT = Online (retained)
21:46:52.704 MQT: cmnd/DevSmartPool/POWER = 
21:46:52.708 MQT: tele/DevSmartPool/INFO1 = {"Info1":{"Module":"ESP32-DevKit","Version":"13.2.0.2(core2)","FallbackTopic":"cmnd/DevSmartPool_fb/","GroupTopic":"cmnd/tasmotas/"}}
21:46:52.723 MQT: tele/DevSmartPool/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"DevSmartPool-0960","IPAddress":"10.1.10.168","IP6Global":"","IP6Local":"fe80::ea9f:6dff:fe09:63c0%st2"}}
21:46:52.738 MQT: tele/DevSmartPool/INFO3 = {"Info3":{"RestartReason":"Vbat power on reset","BootCount":2}}
21:46:52.765 RUL: SYSTEM#BOOT performs "Backlog Discover 0; NPTeleperiod 60;"
21:46:52.811 MQT: stat/DevSmartPool/RESULT = {"Discover":"OFF"}
21:46:53.063 MQT: stat/DevSmartPool/RESULT = {"Command":"Unknown"}
21:46:56.840 MQT: tele/DevSmartPool/STATE = {"Time":"2023-11-15T21:46:56","Uptime":"0T00:00:15","UptimeSec":15,"Heap":139,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":5,"Objects":51},"Wifi":{"AP":1,"SSId":"NetStumbling?","BSSId":"2C:91:AB:AB:72:ED","Channel":1,"Mode":"11n","RSSI":100,"Signal":-8,"LinkCount":1,"Downtime":"0T00:00:10"}}
```

**Reply by sfromis** [reply]
Yeah, that's what you already posted. And my conclusion was that I can't see what's going wrong in your specific setup.

**Reply by alexdelprete** [reply]
Too bad. I'll write a note telling users to check if template is active, and if not, they can issue a `Module 0` command in console.
